### PR TITLE
feat: Add save API for torch-trt compiled models

### PR DESCRIPTION
**Pull request by peri044**
# Description

Add save API for  torch-trt compiled models

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/_compile.py [hunk 1]
```diff
@@ -332,3 +334,73 @@ def convert_method_to_trt_engine(
...
+) -> None:
+    """
+    Save the model to disk in the specified output format.
+    Arguments:
+        module : Compiled Torch-TensorRT module (Options include torch.jit.ScriptModule | torch.export.ExportedProgram | torch.fx.GraphModule)
+        inputs (torch.Tensor): Torch input tensors
+    """
+    module_type = _parse_module_type(module)
+    accepted_formats = {"exported_program", "torchscript"}
+    if inputs and not all(isinstance(input, torch.Tensor) for input in inputs):
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:355]: Switch to `inputs is not None`

File: py/torch_tensorrt/_compile.py [hunk 2]
```diff
@@ -332,3 +334,73 @@ def convert_method_to_trt_engine(
...
+    retrace: bool = False,
+) -> None:
+    """
+    Save the model to disk in the specified output format.
+    Arguments:
+        module : Compiled Torch-TensorRT module (Options include torch.jit.ScriptModule | torch.export.ExportedProgram | torch.fx.GraphModule)
+        inputs (torch.Tensor): Torch input tensors
+    """
+    module_type = _parse_module_type(module)
+    accepted_formats = {"exported_program", "torchscript"}
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:354]: Should the `ep` and `ts` abbreviations be included here?

File: py/torch_tensorrt/_compile.py [hunk 3]
```diff
@@ -332,3 +334,73 @@ def convert_method_to_trt_engine(
...
+            torch.jit.save(module, file_path)
+    elif module_type == _ModuleType.ep:
+        if output_format == "torchscript":
+            raise ValueError(
+                "Provided model is a torch.export.ExportedProgram but the output_format specified is torchscript. Please verify the output_format"
+            )
+        else:
+            torch.export.save(module, file_path)
+    elif module_type == _ModuleType.fx:
+        if not inputs:
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:385]: Consider switching to `inputs is None`

File: py/torch_tensorrt/_compile.py [hunk 4]
```diff
@@ -332,3 +334,73 @@ def convert_method_to_trt_engine(
...
+    output_format: str = "exported_program",
+    inputs: Optional[Sequence[torch.Tensor]] = None,
+    retrace: bool = False,
+) -> None:
+    """
+    Save the model to disk in the specified output format.
+    Arguments:
+        module : Compiled Torch-TensorRT module (Options include torch.jit.ScriptModule | torch.export.ExportedProgram | torch.fx.GraphModule)
+        inputs (torch.Tensor): Torch input tensors
+    """
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:352]: Include a docstring line for the `retrace` argument as well

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 5]
```diff
@@ -55,24 +48,27 @@ def transform(
 
     Returns an inlined torch.fx.GraphModule
     """
+    gm_export = copy.deepcopy(gm)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:51]: Minor question - what is the motivation for not performing this function in-place?

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 6]
```diff
@@ -100,52 +97,69 @@ def lift(gm: torch.fx.GraphModule, graph_signature: Any) -> torch.fx.GraphModule
...
+            else:
+                lift_val = state_dict[node.target]
+
+                input_kind = InputKind.CONSTANT_TENSOR
+
+                # state_dict has these parameters/buffers as torch.Tensors. We override them as torch.nn.Parameter/torch.Tensors respectively.
+                for name, _ in gm.named_parameters():
+                    if node.target == name:
+                        input_kind = InputKind.PARAMETER
+                        state_dict[name] = torch.nn.Parameter(state_dict[name])
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:121]: Should this be `torch.nn.Parameter(state_dict[name], requires_grad=False)`? If it is insignificant, okay to leave as is.

File: tests/py/dynamo/testing_utilities.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 )
 
 DECIMALS_OF_AGREEMENT = 4
+DECIMALS_OF_AGREEMENT_3 = 3
```
Comment (gs-olive) [inline tests/py/dynamo/testing_utilities.py:17]: Is this needed due to a change in this PR?

**Review by gs-olive — COMMENTED** [review]
Overall, looks good. Added some small comments

File: py/torch_tensorrt/_compile.py (hunk 2 above)
Comment (peri044) [reply py/torch_tensorrt/_compile.py:370]: So, our `ir` has `fx` option. In order to not confuse with that, I intentionally did not include the abbreviations. But I don't have a strong preference here to be honest

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 5 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:51]: We deepcopy the graph here because the transformations in this function  as it lifts the graph's parameters/buffers/constants as inputs and the input gm becomes unusable if we perform in-place operation.

File: tests/py/dynamo/testing_utilities.py (hunk 7 above)
Comment (peri044) [reply tests/py/dynamo/testing_utilities.py]: reverted this change. The reason for this change was test_lower_linear wasn't meeting threshold requirement. Reverted to check the latest status on CI.

File: docsrc/user_guide/saving_models.rst [hunk 8]
```diff
@@ -34,50 +33,32 @@ If you set the `output_format="torchscript"`, this will return a `ScriptModule`
...
     import torch_tensorrt
 
     model = MyModel().eval().cuda()
     inputs = [torch.randn((1, 3, 224, 224)).cuda()]
-    # trt_ep is a torch.export.ExportedProgram object
-    trt_ep = torch_tensorrt.compile(model, ir="dynamo", inputs) 
-    torch.export.save(trt_ep, "trt_model.ep")
+    # trt_gm is a torch.fx.GraphModule object
+    trt_gm = torch_tensorrt.compile(model, ir="dynamo", inputs)
+    torchtrt.save(trt_gm, "trt.ts", output_format="torchscript", inputs=inputs)
```
Comment (gs-olive) [inline docsrc/user_guide/saving_models.rst:56]: Replace with `torch_tensorrt.save`, to make the snippet runnable

File: docsrc/user_guide/saving_models.rst [hunk 9]
```diff
@@ -99,3 +80,21 @@ For `ir=ts`, this behavior stays the same in 2.X versions as well.
...
+Here's an example usage
+
+.. code-block:: python
+
+    import torch
+    import torch_tensorrt
+
+    # file_path can be trt.ep or trt.es file obtained via saving the model (refer to the above section)
+    inputs = [torch.randn((1, 3, 224, 224)).cuda()]
+    model = torch_tensorrt.load(<file_path>).module()
```
Comment (gs-olive) [inline docsrc/user_guide/saving_models.rst:99]: Is `"es"` still a valid extension here?

File: py/torch_tensorrt/_compile.py [hunk 10]
```diff
@@ -332,3 +330,98 @@ def convert_method_to_trt_engine(
...
+    except Exception:
+        pass
+
+    try:
+        exp_program = torch.export.load(file_path)
+        return exp_program
+    except Exception:
+        raise ValueError(
+            "The file doesn't correspond to a Torchscript module or ExportedProgram. Please verify the file path."
+        )
```
Comment (gs-olive) [inline py/torch_tensorrt/_compile.py:352]: In both `try/except` clauses, can add `logger.debug("...", exc_info=True)`, to show the exception to the user

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 11]
```diff
@@ -100,52 +100,69 @@ def lift(gm: torch.fx.GraphModule, graph_signature: Any) -> torch.fx.GraphModule
...
-                )
+
+                if isinstance(lift_val, torch.Tensor):
+                    const_placeholder_node.meta["val"] = cast(
+                        FakeTensor,
+                        torch.empty_strided(
+                            tuple(lift_val.shape),
+                            tuple([1] * len(lift_val.shape)),
+                        ),
+                    )
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:149]: I believe the `cast` only has an effect on the linter and has no runtime effect, so this `torch.empty_strided` tensor will be instantiated as memory on the CPU. Is it possible to instead call `FakeTensor` with the desired parameters to avoid instantiation of the tensor?

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 12]
```diff
@@ -100,52 +100,69 @@ def lift(gm: torch.fx.GraphModule, graph_signature: Any) -> torch.fx.GraphModule
...
 
                 node.replace_all_uses_with(const_placeholder_node)
                 gm.graph.erase_node(node)
 
                 # Add these parameters/buffers/constants to the existing graph signature
                 # before user inputs. These specs are looked up in the state_dict during ExportedProgram creation.
+                input_spec_arg = TensorArgument(name=const_placeholder_node.name)
+                if input_kind == InputKind.CUSTOM_OBJ:
+                    input_spec_arg = CustomObjArgument(
+                        name=const_placeholder_node.name, class_fqn=""
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:159]: Are duplicate `class_fqn`s problematic/are they helpful for debugging? If not, ok to leave as is.

**Review by gs-olive — COMMENTED** [review]
Overall, looks great - added a few small comments

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 11 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:149]: > call FakeTensor with the desired parameters to avoid instantiation of the tensor?
I did try this before and this fails (don't have the error log now). I did check with Angela and she recommends `cast` way and they use it internally at a few places. I think cast changes the properties of tensor under the hood for dynamo to treat the node as having meta values but I'm not fully sure.

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 12 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:159]: They are problematic. Recommendation was to leave it empty instead of using `const_placeholder_node.target`
